core/deployment/src/main/java/io/quarkus/deployment/steps/UserProjectRootBuildStep.java

```diff
@@ -0,0 +1,44 @@
+package io.quarkus.deployment.steps;
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import io.quarkus.deployment.annotations.BuildStep;
+import io.quarkus.deployment.builditem.UserProjectRootBuildItem;
+import io.quarkus.deployment.pkg.builditem.OutputTargetBuildItem;
+
+public class UserProjectRootBuildStep {
+
+    /**
+     * Resolves the project directories based on the output target {@link OutputTargetBuildItem}.
+     *
+     * @return a {@link UserProjectRootBuildItem} build item containing the resolved project
+     *         root as {@link java.nio.file.Path}.
+     */
+    @BuildStep
+    UserProjectRootBuildItem rootDir(final OutputTargetBuildItem outputTarget) {
+        Path projectRoot = findProjectRoot(outputTarget.getOutputDirectory());
+        if (projectRoot != null) {
+            return new UserProjectRootBuildItem(projectRoot);
+        }
+        return null;
+    }
+
+    private static Path findProjectRoot(Path outputDirectory) {
+        Path currentPath = outputDirectory;
+        do {
+            if (Files.exists(currentPath.resolve(Paths.get("src", "main")))
+                    || Files.exists(currentPath.resolve(Paths.get("config", "application.properties")))
```

Comment:
Why is /config checked?  Shouldn't it just be `/src/main/resources/application.properties` etc?

Reply:
It would probably be better to call it a "module dir" rather than a "project root", since this code is not looking for the root directory of a possibly multimodule project.

There could be a case when there won't be a module directory, e.g. when we build an application from a JAR, there is no workspace around.

To avoid relying on the assumption about a certain directory structure and specific files being present, I think we could explore a possibility of adding a method to `CurateOutcomeBuildItem` such as `getApplicationModuleDir()` that would derive it from the information available in `CurateOutcomeBuildItem.getApplicationModel().getAppArtifact()`. It could check whether `getWorkspaceModule()` is not null and if it isn't it would call `getModuleDir()` on it.
If it is null, it could use the current directory, I guess, or return null, depending on what this directory is used for.

Reply:
@aloubyansky we are open to anything you suggest as like @ia3andy mentioned we have this same cut and pasted code in at least 5 or 6 extensions already 😄

Reply:
It depends if some of you wanted to do that or prefer me opening another PR and reviewing that.

Reply:
I think we are both OK with you opening a new PR for us to review?